keep-ui/app/rules/AIGenRules.tsx

```diff
@@ -0,0 +1,318 @@
+import { useSession } from "next-auth/react";
+import { getApiURL } from "utils/apiUrl";
+import useSWR, { mutate } from "swr";
+import Loading from "app/loading";
+
+
+
+const columnHelper = createColumnHelper<AIGeneratedRule>();
+
+interface SortableHeaderCellProps {
+  header: Header<AIGeneratedRule, unknown>;
+  children: React.ReactNode;
+}
+
+const SortableHeaderCell: React.FC<SortableHeaderCellProps> = ({
+  header,
+  children,
+}) => {
+  const { column } = header;
+
+  return (
+    <TableHeaderCell
+      className={`relative ${column.getIsPinned() ? "" : "hover:bg-slate-100"} group`}
+    >
+      <div className="flex items-center">
+        {children} {/* Column name or text */}
+        {column.getCanSort() && (
+          <>
+            {/* Custom styled vertical line separator */}
+            <div className="w-px h-5 mx-2 bg-gray-400"></div>
+            <Icon
+              className="cursor-pointer"
+              size="xs"
+              color="neutral"
+              onClick={(event) => {
+                event.stopPropagation();
+                const toggleSorting = header.column.getToggleSortingHandler();
+                if (toggleSorting) toggleSorting(event);
+              }}
+              tooltip={
+                column.getNextSortingOrder() === "asc"
+                  ? "Sort ascending"
+                  : column.getNextSortingOrder() === "desc"
+                  ? "Sort descending"
+                  : "Clear sort"
+              }
+              icon={column.getIsSorted() ? (
+                column.getIsSorted() === "asc" ? FaArrowDown : FaArrowUp
+              ) : (
+                FaArrowRight
+              )}
+            />
+          </>
+        )}
+      </div>
+    </TableHeaderCell>
+  );
+};
+
+export const AIGenRules: React.FC = () => {
+  const {triggerGenRules } = useGenRules();
+  const { data: session } = useSession();
+  const [sorting, setSorting] = React.useState<SortingState>([]);
+
+  const [isLoadingRules, setIsLoadingRules] = useState(true);
+  const [generatedRules, setGeneratedRules] = useState<any>([]);
+
+  const [loadingRows, setLoadingRows] = useState<{ [key: string]: boolean }>({});
+  const [successRows, setSuccessRows] = useState<{ [key: string]: boolean }>({});
+
+  const mutateAIGeneratedRules = (rules: AIGeneratedRule[]) => {
+    setGeneratedRules(rules);
+    setIsLoadingRules(false);
+  };
+
+  const { data:serverGenRules } = useRulePusherUpdates();
+
+  useEffect(() => {
+    console.log("serverGenRules", serverGenRules);
+
+    if (Array.isArray(serverGenRules) && (0 === serverGenRules.length)) {
+      return;
+    }
+    
+    mutateAIGeneratedRules(serverGenRules);
+  }, [serverGenRules]);
+
+  const handleGenerateMoreRules = () => {
+    setIsLoadingRules(true);
+    triggerGenRules();
+  };
+
+  const handleAddRule = async (rule: AIGeneratedRule) => {
```

Comment:
Let's move `addRule()` function to `useRules()` hook to isolate data logic to the hook and this component to mostly UI.